src/AddressSpace.cc

```diff
@@ -881,22 +884,33 @@ void AddressSpace::unmap_internal(remote_ptr<void> addr, ssize_t num_bytes) {
 
     Mapping m = move(mm);
     mem.erase(m.map);
+
+    // Also unmap any corresponding mapping from the local address space
+    if (m.local_addr) {
+      uint8_t* start_addr =
+          m.local_addr + (max(m.map.start(), rem.start()) - m.map.start());
+      uint8_t* stop_addr =
+          m.local_addr + (min(m.map.end(), rem.end()) - m.map.end());
+      munmap(start_addr, stop_addr - start_addr);
+    }
```

Comment:
So I think we shouldn't touch any local mappings in AddressSpace and have the caller take responsibility for it. If it helps, we could add the ability to associate a callback with a mapped memory range that gets called when there are no more mapped pages.